### DISCUSSION TITLE
Tab expand environment variable

### DISCUSSION THREAD

**Discussion by MCoooo** [nushell/nushell#6632] [Q&A]
Hi All,

I'm trying to get to grips with nu and wish to use it as a daily driver on Windows.

A piece of functionality I'm missing (\haven't set up?) is tab expanding environment variables on the command line to navigate file paths.

An example would be $env.APPDATA. Rather than being able to expand then continue navigating sub directories, I have to select return for appdata to resolve then enter the sub path as a second path.

Any guidance would be much appreciated.

Cheers.

**Comment by fdncred** [comment]
We don't offer expansion of env vars on the command line atm. We'd probably welcome a keybinding or some such thing to accomplish it. It's kind of the same idea as `abbr` abbreviations.

**Reply by MCoooo** [reply]
Thank you. I will have to do some reading up!

**Comment by moonstruckmoth** [comment]
I've been using nushell for a couple of weeks and variable expansion is something I used a lot in zsh. 
And which I missed a lot in nushell.

So for now I'm using the following (very rough) workaround

I capture the environment variables into a json file with the following
> $env | transpose name value | where (($it.value | describe) == string and ($it.value != '')) | to json | save ~/tmp/env.json

I've added the following to the $env.config.menus (in config.nu)

        {
          name: env_vars_menu
          only_buffer_difference: true
          marker: "👀 "
          type: {
            layout: list
            page_size: 10
          }
          style: {
            text: green
            selected_text: green_reverse
            description_text: yellow
          }
          source: { |buffer, position|
            open ~/tmp/env.json
            | where ($buffer | str length) >= 2 and name =~ ^($buffer)
            | each { |it| {value: $it.value description: $it.name }}
          }
        }

And the following to $env.config.keybindings

        {
          name: env_vars
          modifier: control
          keycode: char_g
          mode: [emacs, vi_normal, vi_insert]
          event: [
            { send: menu name: env_vars_menu }
          ]
        }

My keybind is ctrl-g, so how it works is,
pressing ctrl-g and typing the variable name completes it.

So for eg.
ctrl-g + TZ + enter will replace it with value of TZ env variable

Need to test more though

**Reply by moonstruckmoth** [reply]
Updated code to use regex for matching and trigger matches when input length >= 2 characters.
(Without the input length conditions, ctrl-g dumps all values.)

For eg with following variables set
LESSCHARSET="latin1"
LESSOPEN="| lesspipe.sh %s"
LESS="-i  -e -M -X -F -R -P%t?f%f :stdin .?pb%pb\%:?lbLine %lb:?bbByte %bb:-..."

Now type ctrl-g + LESS will display all three matches

![image](https://github.com/user-attachments/assets/8f100bcd-21c9-40b9-8625-1320b470b8dc)

**Reply by mkatychev** [reply]
Thanks for the example @moonstruckmoth  I modified your logic into a key command that makes `CTRL + g` expand `$env` variables to their string value (if they're `path` or `string`) without need for file:

```nushell
def fill-env-var [] {
    let current_prompt = (commandline)
    let var_idx = $current_prompt | str index-of "$env." --end
    let head = match $var_idx {
        # no '$env.' was found
        -1 => return
        0 => ""
        $idx => ($current_prompt | str substring ..($idx - 1))
    }
    # 5 is the offset for f in '$env.foo`
    mut var = $current_prompt | str substring ($var_idx + 5)..
    if ($var | str length) == 0 {
        return
    }
    let tail = match ($var | str index-of " ") {
        -1 => ""
        $idx => {
            let tail = $var | str substring $idx..
            $var = $var | str substring ..($idx - 1)
            $tail
        }
    }
    let var_path = $var | split words | into cell-path
    let val = $env | get --optional $var_path
    match ($val | describe) {
       "path" | "string" =>  {}
       _ => return
    }
    let output = [$head $val $tail] | str join

    commandline edit --replace $output
    commandline set-cursor --end
}

$env.config.keybindings ++= [
    {
        name: env_vars
        modifier: control
        keycode: char_g
        mode: [emacs vi_insert vi_normal]
        event: [
            {send: executehostcommand cmd: fill-env-var}
        ]
    }
]
```

https://gist.github.com/mkatychev/86321b557259e41a742774b5a1a60ecf
